File: matiasmasso/Mat.NET/MatHelperCFwk/DocfileHelper.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using MatHelperStd;
using Microsoft.VisualBasic;
using DTO;
using LegacyHelper;
using System.Resources;
using System.Security.Policy;
using System.Security.Cryptography;

namespace MatHelperCFwk
{


    public class DocfileHelper
    {
        public static byte[] ZipThumbnail(byte[] oByteArray)
        {
            byte[] retval = null;
            List<string> sFilenames = ZipHelper.Filenames(oByteArray);
            if (sFilenames.Count > 0)
            {
                var oFiles = ZipHelper.Extract(oByteArray);
                if (oFiles.Count > 0)
                {
                    List<Exception> exs = new List<Exception>();
                    var oFirstDocfile = DocfileHelper.Factory(exs, oFiles.First().ByteArray);
                    retval = oFirstDocfile.Thumbnail;
                }
            }
            return retval;
        }

        public static List<DTODocFile> Factory(List<byte[]> oFiles, List<Exception> exs)
        {
            List<DTODocFile> retval = new List<DTODocFile>();
            foreach (byte[] oFile in oFiles)
            {
                DTODocFile oDocFile = DocfileHelper.Factory(exs, oFile, MimeCods.NotSet);
                retval.Add(oDocFile);
            }
            return retval;
        }

        public static DTODocFile Factory(List<Exception> exs, byte[] oByteArray, MimeCods oMime = MimeCods.NotSet)
        {
            DTODocFile retval = new DTODocFile();
            LoadFromStream(exs, ref retval, oByteArray, oMime);
            return retval;
        }

        public static DTODocFile LoadFromString(string src)
        {
            var oByteArray = System.Text.Encoding.UTF8.GetBytes(src);
            var retval = new DTODocFile()
            {
                Hash = MatHelperCFwk.ASINHelper.GenerateRandomASIN(),
                Sha256 = CryptoHelper.Sha256(oByteArray),
                Thumbnail = TextThumbnail(src),
                FchCreated = DTO.GlobalVariables.Now(),
                Stream = oByteArray,
                Mime = MimeCods.Txt,
                Length = oByteArray.Length
            };
            return retval;
        }



        public static byte[] TextThumbnail(string src)
        {
            System.Drawing.Bitmap oThumbnail = new System.Drawing.Bitmap(DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT);
            Graphics oGraphics = Graphics.FromImage(oThumbnail);
            oGraphics.FillRectangle(Brushes.White, 0, 0, oThumbnail.Width, oThumbnail.Height);
            var oFont = new Font("Helvetica", 10, FontStyle.Regular);
            int padding = 10;
            System.Drawing.RectangleF oRectangleF = new System.Drawing.RectangleF(padding, padding, DTODocFile.THUMB_WIDTH - 2 * padding, DTODocFile.THUMB_HEIGHT - 2 * padding);
            StringFormat sf = new StringFormat();
            oGraphics.DrawString(src, oFont, Brushes.Black, oRectangleF, sf);
            byte[] retval = null;
            using (System.IO.MemoryStream ms = new System.IO.MemoryStream())
            {
                oThumbnail.Save(ms, System.Drawing.Imaging.ImageFormat.Jpeg);
                retval = ms.ToArray();
            }
            return retval;
        }

        public static DTODocFile Factory(string sFilename, List<Exception> exs)
        {
            DTODocFile retval = null/* TODO Change to default(_) if this is not a reference type */;
            byte[] oByteArray = null;
            if (FileSystemHelper.GetStreamFromFile(sFilename, ref oByteArray, ref exs))
            {
                var oMime = MimeHelper.GetMimeFromExtension(sFilename);
                retval = Factory(exs, oByteArray, oMime);
                retval.Filename = sFilename;
            }
            return retval;
        }

        public static bool LoadFromStream(byte[] oByteArray, ref DTODocFile oDocFile, List<Exception> exs, string sFilename = "")
        {
            bool retval = false;
            if (oByteArray == null)
                exs.Add(new Exception("ByteArray buit al generar el DocFile"));
            else
            {
                oDocFile = new DTODocFile();
                {
                    var withBlock = oDocFile;
                    withBlock.Hash = MatHelperCFwk.ASINHelper.GenerateRandomASIN();
                    withBlock.Sha256 = CryptoHelper.Sha256(oByteArray);
                    withBlock.Stream = oByteArray;
                    withBlock.Length = oByteArray.Length;
                    if (!string.IsNullOrEmpty(sFilename))
                    {
                        withBlock.Filename = sFilename;
                        withBlock.Mime = MimeHelper.GetMimeFromExtension(sFilename);
                    }
                    if (withBlock.Mime == MimeCods.NotSet)
                        withBlock.Mime = MimeHelper.GuessMime(oByteArray);
                    if (LoadMimeDetails(oDocFile, exs))
                        retval = true;
                }
            }
            return retval;
        }

        public static bool LoadMimeDetails(DTODocFile oDocFile, List<Exception> exs)
        {
            bool retval = true;
            {
                var withBlock = oDocFile;
                switch (withBlock.Mime)
                {
                    case MimeCods.Pdf:
                        {
                            retval = LoadPdf(ref oDocFile, exs);
                            break;
                        }

                    case MimeCods.Rtf:
                        {
                            break;
                        }

                    case MimeCods.Zip:
                        {
                            var oFiles = ZipHelper.Extract(withBlock.Stream);
                            if (oFiles.Count > 0)
                            {
                                var oFirstDocfile = DocfileHelper.Factory(exs, oFiles.First().ByteArray);
                                withBlock.Thumbnail = oFirstDocfile.Thumbnail;
                            }

                            break;
                        }

                    case MimeCods.Jpg:
                    case MimeCods.Gif:
                    case MimeCods.Png:
                    case MimeCods.Bmp:
                    case MimeCods.Tif:
                    case MimeCods.Tiff:
                        {
                            System.IO.MemoryStream ms = new System.IO.MemoryStream(withBlock.Stream);
                            var oImg = System.Drawing.Image.FromStream(ms);
                            withBlock.Size = new Size(oImg.Width, oImg.Height);
                            withBlock.VRes = (int)oImg.VerticalResolution;
                            withBlock.HRes = (int)oImg.HorizontalResolution;
                            System.Drawing.Image oThumbnail = ImageHelper.GetThumbnailToFit((Bitmap)oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT);
                            using (System.IO.MemoryStream ms2 = new System.IO.MemoryStream())
                            {
                                oThumbnail.Save(ms2, System.Drawing.Imaging.ImageFormat.Jpeg);
                                withBlock.Thumbnail = ms2.ToArray();
                            }

                            break;
                        }

                    case MimeCods.Ppt:
                    case MimeCods.Pptx:
                        {
                            break;
                        }

                    case MimeCods.Docx:
                        {
                            break;
                        }

                    case MimeCods.Txt:
                        {
                            string src = System.Text.ASCIIEncoding.ASCII.GetString(oDocFile.Stream);
                            var oThumbnail = ImageHelper.GetThumbnailFromText(src, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT);
                            withBlock.Thumbnail = oThumbnail.Bytes();
                            break;
                        }

                    case MimeCods.Mov:
                    case MimeCods.Mp4:
                    case MimeCods.Mpg:
                        {
                            Image oImg = new Bitmap(
    System.Reflection.Assembly.GetEntryAssembly().
      GetManifestResourceStream("MatHelperCFwk.Resources.video.jpg"));
                            withBlock.Thumbnail = ImageHelper.GetThumbnailToFit((Bitmap)oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes();
                            break;
                        }

                    default:
                        {
                            Image oImg = new Bitmap(
    System.Reflection.Assembly.GetEntryAssembly().
      GetManifestResourceStream("MatHelperCFwk.Resources.cartadeajuste.png"));
                            withBlock.Thumbnail = ImageHelper.GetThumbnailToFit((Bitmap)oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes();
                            break;
                        }
                }
            }
            return retval;
        }


        public static bool LoadFromStream(List<Exception> exs, ref DTODocFile oDocfile, byte[] oByteArray, MimeCods oMime = MimeCods.NotSet)
        {
            try
            {
                {
                    var withBlock = oDocfile;
                    withBlock.Hash = MatHelperCFwk.ASINHelper.GenerateRandomASIN();
                    withBlock.Sha256 = CryptoHelper.Sha256(oByteArray);
                    withBlock.FchCreated = DTO.GlobalVariables.Now();
                    withBlock.Stream = oByteArray;
                    if (oMime == MimeCods.NotSet)
                        withBlock.Mime = MimeHelper.GuessMime(oByteArray);
                    else
                        withBlock.Mime = oMime;
                    withBlock.Length = oByteArray.Length;

                    switch (withBlock.Mime)
                    {
                        case MimeCods.Pdf:
                            {
                                System.IO.MemoryStream oStream = new System.IO.MemoryStream(oByteArray);
                                var oPdf = GhostScriptHelper.Rasterize(exs, oStream);
                                withBlock.Thumbnail = oPdf.Thumbnail;
                                withBlock.Pags = oPdf.PageCount;
                                withBlock.Size = new Size(oPdf.Width, oPdf.Height);
                                break;
                            }

                        case MimeCods.Rtf:
                            {
                                break;
                            }

                        case MimeCods.Zip:
                            {
                                withBlock.Thumbnail = DocfileHelper.ZipThumbnail(withBlock.Stream);
                                break;
                            }

                        case MimeCods.Jpg:
                        case MimeCods.Gif:
                        case MimeCods.Png:
                        case MimeCods.Bmp:
                        case MimeCods.Tif:
                        case MimeCods.Tiff:
                            {
                                System.IO.MemoryStream ms = new System.IO.MemoryStream(withBlock.Stream);
                                var oImg = Image.FromStream(ms);
                                withBlock.Size = new Size(oImg.Width, oImg.Height);
                                withBlock.VRes = (int)oImg.VerticalResolution;
                                withBlock.HRes = (int)oImg.HorizontalResolution;
                                withBlock.Thumbnail = ImageHelper.GetThumbnailToFit((Bitmap)oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes();
                                break;
                            }

                        case MimeCods.Ppt:
                        case MimeCods.Pptx:
                            {
                                break;
                            }

                        case MimeCods.Docx:
                            {
                                withBlock.Thumbnail = WordHelper.GetImgFromWordFirstPage(withBlock.Stream, exs);
                                break;
                            }

                        case MimeCods.Xlsx:
                            {
                                int iExcelRows = 0;
                                int iExcelCols = 0;
                                // .Thumbnail = MatHelper.Excel.GetImgFromExcelFirstPage(_DocFile.Stream, iExcelCols, iExcelRows)
                                // .Thumbnail = My.Resources.Excel_Big
                                withBlock.Thumbnail = MatHelper.Excel.Rasterizer.GetExcelThumbnail(withBlock.Stream, exs);
                                withBlock.Size = new Size(iExcelCols, iExcelRows);
                                break;
                            }

                        case MimeCods.Txt:
                            {
                                string src = System.Text.ASCIIEncoding.ASCII.GetString(oByteArray);
                                break;
                            }

                        case MimeCods.Mov:
                        case MimeCods.Mp4:
                            {
                                Image oImg = new Bitmap(
                                    System.Reflection.Assembly.GetEntryAssembly().
                                      GetManifestResourceStream("MatHelperCFwk.Resources.video.jpg"));
                                withBlock.Thumbnail = ImageHelper.GetThumbnailToFit((Bitmap)oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes();
                                break;
                            }
                    }
                }
            }
            catch (SystemException e)
            {
                exs.Add(e);
            }
            return exs.Count == 0;
        }


        public static bool LoadPdf(ref DTODocFile oDocFile, List<Exception> exs)
        {
            bool retval = false;
            GhostScriptHelper.Pdf oPdf = null/* TODO Change to default(_) if this is not a reference type */;
            if (Load(ref oPdf, oDocFile.Stream, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT, exs))
            {
                {
                    var withBlock = oDocFile;
                    withBlock.Size = new Size(oPdf.Width, oPdf.Height);
                    withBlock.Pags = oPdf.PageCount;
                    withBlock.Thumbnail = oPdf.Thumbnail;
                }
                retval = true;
            }
            return retval;
        }


        public static byte[] GetPdfThumbnail(byte[] oFileBytes, int iMaxWidth, int iMaxHeight)
        {
            GhostScriptHelper.Pdf oPdf = null/* TODO Change to default(_) if this is not a reference type */;
            List<Exception> exs = new List<Exception>();
            byte[] retval = null;
            if (Load(ref oPdf, oFileBytes, iMaxWidth, iMaxHeight, exs))
                retval = oPdf.Thumbnail;
            return retval;
        }

        public static byte[] GetThumbnail(System.IO.MemoryStream oStream, int iMaxWidth, int iMaxHeight)
        {
            GhostScriptHelper.Pdf oPdf = null/* TODO Change to default(_) if this is not a reference type */;
            List<Exception> exs = new List<Exception>();
            byte[] retval = null;
            if (Load(ref oPdf, oStream, iMaxWidth, iMaxHeight, exs))
                retval = oPdf.Thumbnail;
            return retval;
        }

        public static bool Load(string sFilename, ref byte[] oByteArray, int maxWidth, int MaxHeight, List<Exception> exs)
        {
            bool retval = false;
            if (FileSystemHelper.GetStreamFromFile(sFilename, ref oByteArray, ref exs))
            {
                GhostScriptHelper.Pdf oPdf = new GhostScriptHelper.Pdf();
                retval = Load(ref oPdf, oByteArray, maxWidth, MaxHeight, exs);
            }
            return retval;
        }

        public static bool Load(ref GhostScriptHelper.Pdf oPdf, byte[] oByteArray, int maxWidth, int maxHeight, List<Exception> exs)
        {
            System.IO.MemoryStream oStream = new System.IO.MemoryStream(oByteArray);
            bool retval = Load(ref oPdf, oStream, maxWidth, maxHeight, exs);
            return retval;
        }
        public static bool LoadToFit(ref GhostScriptHelper.Pdf oPdf, byte[] oByteArray, int maxWidth, int maxHeight, List<Exception> exs)
        {
            System.IO.MemoryStream oStream = new System.IO.MemoryStream(oByteArray);
            bool retval = LoadToFit(ref oPdf, oStream, maxWidth, maxHeight, exs);
            return retval;
        }

        public static bool Load(ref GhostScriptHelper.Pdf oPdf, string sFilename, int maxWidth, int maxHeight, List<Exception> exs)
        {
            System.IO.Stream oStream = new System.IO.FileStream(sFilename, System.IO.FileMode.Open, System.IO.FileAccess.Read);
            bool retval = Load(ref oPdf, oStream, maxWidth, maxHeight, exs);
            return retval;
        }

        public static bool Load(ref GhostScriptHelper.Pdf oPdf, System.IO.Stream oStream, int maxWidth, int maxHeight, List<Exception> exs)
        {
            bool retval = GhostScriptHelper.Rasterize(oStream, ref oPdf, exs);
            if (retval)
            {
                if (maxWidth == 0 & maxHeight == 0)
                    oPdf.Thumbnail = oPdf.Portrait;
                else
                    oPdf.Thumbnail = ImageHelper.GetThumbnailToFill((Bitmap)ImageHelper.FromBytes(oPdf.Portrait), maxWidth, maxHeight).Bytes();
            }
            return retval;
        }

        public static bool LoadToFit(ref GhostScriptHelper.Pdf oPdf, System.IO.Stream oStream, int maxWidth, int maxHeight, List<Exception> exs)
        {
            bool retval = GhostScriptHelper.Rasterize(oStream, ref oPdf, exs);
            if (retval)
            {
                if (maxWidth == 0 & maxHeight == 0)
                    oPdf.Thumbnail = oPdf.Portrait;
                else
                {
                    var oPortrait = ImageHelper.FromBytes(oPdf.Portrait);
                    oPdf.Thumbnail = ImageHelper.GetThumbnailToFit((Bitmap)oPortrait, maxWidth, maxHeight).Bytes();
                }
            }
            return retval;
        }

        public static byte[] ThumbnailPreview(DTODocFile oDocFile)
        {
            byte[] retval = null;
            if (oDocFile != null)
            {
                switch (oDocFile.Mime)
                {
                    case MimeCods.Bmp:
                    case MimeCods.Gif:
                    case MimeCods.Jpg:
                    case MimeCods.Png:
                    case MimeCods.Tif:
                    case MimeCods.Tiff:
                        {
                            var oImg = ImageHelper.FromBytes(oDocFile.Thumbnail);
                            retval = ImageHelper.GetThumbnailToFill((Bitmap)oImg, 100, 100).Bytes();
                            break;
                        }

                    case MimeCods.Pdf:
                        {
                            break;
                        }
                }
            }
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/LegacyHelper/DocfileHelper.vb

Public Class DocfileHelper


    Shared Function ZipThumbnail(oByteArray As Byte()) As Byte()
        Dim retval As Byte() = Nothing
        Dim sFilenames As List(Of String) = LegacyHelper.ZipHelper.Filenames(oByteArray)
        If sFilenames.Count > 0 Then
            Dim oFiles = LegacyHelper.ZipHelper.Extract(oByteArray)
            If oFiles.Count > 0 Then
                Dim exs As New List(Of Exception)
                Dim oFirstDocfile = DocfileHelper.Factory(exs, oFiles.First.ByteArray)
                retval = oFirstDocfile.Thumbnail
            End If
        End If
        Return retval
    End Function

    Shared Function Factory(oFiles As List(Of Byte()), exs As List(Of Exception)) As List(Of DTODocFile)
        Dim retval As New List(Of DTODocFile)
        For Each oFile As Byte() In oFiles
            Dim oDocFile As DTODocFile = DocfileHelper.Factory(exs, oFile, MimeCods.NotSet)
            retval.Add(oDocFile)
        Next
        Return retval
    End Function

    Shared Function Factory(exs As List(Of Exception), oByteArray As Byte(), Optional oMime As MimeCods = MimeCods.NotSet) As DTODocFile
        Dim retval As New DTODocFile
        LoadFromStream(exs, retval, oByteArray, oMime)
        Return retval
    End Function

    Shared Function LoadFromString(src As String) As DTODocFile
        Dim oByteArray = Text.Encoding.UTF8.GetBytes(src)
        Dim retval As New DTODocFile()
        With retval
            .Hash = ASINHelper.GenerateRandomASIN()
            .Sha256 = CryptoHelper.Sha256(oByteArray)
            .Thumbnail = TextThumbnail(src)
            .FchCreated = DTO.GlobalVariables.Now()
            .Stream = oByteArray
            .Mime = MimeCods.Txt
            .Length = oByteArray.Length
        End With
        Return retval
    End Function

    Shared Function TextThumbnail(src As String) As Byte()
        Dim oThumbnail As New System.Drawing.Bitmap(DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT)
        Dim oGraphics As Graphics = Graphics.FromImage(oThumbnail)
        oGraphics.FillRectangle(Brushes.White, 0, 0, oThumbnail.Width, oThumbnail.Height)
        Dim oFont = New Font("Helvetica", 10, FontStyle.Regular)
        Dim padding As Integer = 10
        Dim oRectangleF As New System.Drawing.RectangleF(padding, padding, DTODocFile.THUMB_WIDTH - 2 * padding, DTODocFile.THUMB_HEIGHT - 2 * padding)
        Dim sf As New StringFormat()
        oGraphics.DrawString(src, oFont, Brushes.Black, oRectangleF, sf)
        Dim retval As Byte() = Nothing
        Using ms As New IO.MemoryStream()
            oThumbnail.Save(ms, Imaging.ImageFormat.Jpeg)
            retval = ms.ToArray()
        End Using
        Return retval
    End Function

    Shared Function Factory(sFilename As String, exs As List(Of Exception)) As DTODocFile
        Dim retval As DTODocFile = Nothing
        Dim oByteArray As Byte() = Nothing
        If FileSystemHelper.GetStreamFromFile(sFilename, oByteArray, exs) Then
            Dim oMime = MimeHelper.GetMimeFromExtension(sFilename)
            retval = Factory(exs, oByteArray, oMime)
            retval.filename = sFilename
        End If
        Return retval
    End Function

    Shared Function LoadFromStream(oByteArray As Byte(), ByRef oDocFile As DTODocFile, exs As List(Of Exception), Optional sFilename As String = "") As Boolean
        Dim retval As Boolean
        If oByteArray Is Nothing Then
            exs.Add(New Exception("ByteArray buit al generar el DocFile"))
        Else
            oDocFile = New DTODocFile
            With oDocFile
                .Hash = AsinHelper.GenerateRandomASIN()
                .Sha256 = CryptoHelper.Sha256(oByteArray)
                .Stream = oByteArray
                .length = oByteArray.Length
                If sFilename > "" Then
                    .Filename = sFilename
                    .Mime = MimeHelper.GetMimeFromExtension(sFilename)
                End If
                If .mime = MimeCods.NotSet Then
                    .mime = MimeHelper.GuessMime(oByteArray)
                End If
                If LoadMimeDetails(oDocFile, exs) Then
                    retval = True
                End If
            End With
        End If
        Return retval
    End Function

    Shared Function LoadMimeDetails(oDocFile As DTODocFile, exs As List(Of Exception)) As Boolean
        Dim retval As Boolean = True
        With oDocFile
            Select Case .mime
                Case MimeCods.Pdf
                    retval = LoadPdf(oDocFile, exs)
                Case MimeCods.Rtf
                    '.Thumbnail = LegacyHelper.ImageHelper.Converter(My.Resources.word)
                Case MimeCods.Zip
                    Dim oFiles = ZipHelper.Extract(.Stream)
                    If oFiles.Count > 0 Then
                        Dim oFirstDocfile = DocfileHelper.Factory(exs, oFiles.First.ByteArray)
                        .Thumbnail = oFirstDocfile.Thumbnail
                    End If
                Case MimeCods.Jpg, MimeCods.Gif, MimeCods.Png, MimeCods.Bmp, MimeCods.Tif, MimeCods.Tiff
                    Dim ms As New IO.MemoryStream(.Stream)
                    Dim oImg = System.Drawing.Image.FromStream(ms)
                    .Size = New Size(oImg.Width, oImg.Height)
                    .VRes = oImg.VerticalResolution
                    .HRes = oImg.HorizontalResolution
                    Dim oThumbnail As System.Drawing.Image = LegacyHelper.ImageHelper.GetThumbnailToFit(oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT)
                    Using ms2 As New System.IO.MemoryStream
                        oThumbnail.Save(ms2, Imaging.ImageFormat.Jpeg)
                        .Thumbnail = ms2.ToArray()
                    End Using
                Case MimeCods.Ppt, MimeCods.Pptx
                Case MimeCods.Docx
                Case MimeCods.Txt
                    Dim src As String = System.Text.ASCIIEncoding.ASCII.GetString(oDocFile.Stream)
                    Dim oThumbnail = LegacyHelper.ImageHelper.GetThumbnailFromText(src, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT)
                    .Thumbnail = oThumbnail.Bytes()

                Case MimeCods.Mov, MimeCods.Mp4, MimeCods.Mpg
                    Dim oImg As Image = My.Resources.video
                    .Thumbnail = ImageHelper.GetThumbnailToFit(oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes()
                Case Else
                    Dim oImg As Image = My.Resources.cartadeajuste
                    .Thumbnail = ImageHelper.GetThumbnailToFit(oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes()

            End Select
        End With
        Return retval
    End Function


    Shared Function LoadFromStream(exs As List(Of Exception), ByRef oDocfile As DTODocFile, oByteArray As Byte(), Optional oMime As MimeCods = MimeCods.NotSet) As Boolean
        Try
            With oDocfile
                .Hash = AsinHelper.GenerateRandomASIN()
                .Sha256 = CryptoHelper.Sha256(oByteArray)
                .FchCreated = DTO.GlobalVariables.Now()
                .Stream = oByteArray
                If oMime = MimeCods.NotSet Then
                    .mime = MimeHelper.GuessMime(oByteArray)
                Else
                    .mime = oMime
                End If
                .length = oByteArray.Length

                Select Case .mime
                    Case MimeCods.Pdf
                        Dim oStream As New System.IO.MemoryStream(oByteArray)
                        Dim oPdf = GhostScriptHelper.Rasterize(exs, oStream)
                        .Thumbnail = oPdf.Thumbnail
                        .pags = oPdf.PageCount
                        .Size = New Size(oPdf.Width, oPdf.Height)
                    Case MimeCods.Rtf
                        '.Thumbnail = LegacyHelper.ImageHelper.Converter(My.Resources.word)
                    Case MimeCods.Zip
                        .Thumbnail = DocfileHelper.ZipThumbnail(.Stream)
                    Case MimeCods.Jpg, MimeCods.Gif, MimeCods.Png, MimeCods.Bmp, MimeCods.Tif, MimeCods.Tiff
                        Dim ms As New IO.MemoryStream(.Stream)
                        Dim oImg = Image.FromStream(ms)
                        .Size = New Size(oImg.Width, oImg.Height)
                        .VRes = oImg.VerticalResolution
                        .HRes = oImg.HorizontalResolution
                        .Thumbnail = ImageHelper.GetThumbnailToFit(oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes()
                    Case MimeCods.Ppt, MimeCods.Pptx
                    Case MimeCods.Docx
                        .Thumbnail = WordHelper.GetImgFromWordFirstPage(.Stream, exs)
                    Case MimeCods.Xlsx
                        Dim iExcelRows As Integer = 0
                        Dim iExcelCols As Integer = 0
                        '.Thumbnail = MatHelper.Excel.GetImgFromExcelFirstPage(_DocFile.Stream, iExcelCols, iExcelRows)
                        '.Thumbnail = My.Resources.Excel_Big
                        .Thumbnail = ExcelThumbnail.GetExcelThumbnail(.Stream, exs)
                        .Size = New Size(iExcelCols, iExcelRows)
                    Case MimeCods.Txt
                        Dim src As String = System.Text.ASCIIEncoding.ASCII.GetString(oByteArray)
                        '.Thumbnail = ImageHelper.GetThumbnailFromText(src, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT)
                    Case MimeCods.Mov, MimeCods.Mp4
                        Dim oImg As Image = My.Resources.video
                        .Thumbnail = ImageHelper.GetThumbnailToFit(oImg, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT).Bytes()
                End Select

            End With

        Catch e As SystemException
            exs.Add(e)
        End Try
        Return exs.Count = 0
    End Function


    Shared Function LoadPdf(ByRef oDocFile As DTODocFile, exs As List(Of Exception)) As Boolean
        Dim retval As Boolean
        Dim oPdf As GhostScriptHelper.Pdf = Nothing
        If Load(oPdf, oDocFile.Stream, DTODocFile.THUMB_WIDTH, DTODocFile.THUMB_HEIGHT, exs) Then
            With oDocFile
                .Size = New Size(oPdf.Width, oPdf.Height)
                .pags = oPdf.PageCount
                .Thumbnail = oPdf.Thumbnail
            End With
            retval = True
        End If
        Return retval
    End Function


    Shared Function GetPdfThumbnail(oFileBytes As Byte(), iMaxWidth As Integer, iMaxHeight As Integer) As Byte()
        Dim oPdf As GhostScriptHelper.Pdf = Nothing
        Dim exs As New List(Of Exception)
        Dim retval As Byte() = Nothing
        If Load(oPdf, oFileBytes, iMaxWidth, iMaxHeight, exs) Then
            retval = oPdf.Thumbnail
        End If
        Return retval
    End Function

    Shared Function GetThumbnail(oStream As System.IO.MemoryStream, iMaxWidth As Integer, iMaxHeight As Integer) As Byte()
        Dim oPdf As GhostScriptHelper.Pdf = Nothing
        Dim exs As New List(Of Exception)
        Dim retval As Byte() = Nothing
        If Load(oPdf, oStream, iMaxWidth, iMaxHeight, exs) Then
            retval = oPdf.Thumbnail
        End If
        Return retval
    End Function

    Shared Function Load(sFilename As String, oByteArray As Byte(), maxWidth As Integer, MaxHeight As Integer, exs As List(Of Exception)) As Boolean
        Dim retval As Boolean
        If FileSystemHelper.GetStreamFromFile(sFilename, oByteArray, exs) Then
            Dim oPdf As New GhostScriptHelper.Pdf
            retval = Load(oPdf, oByteArray, maxWidth, MaxHeight, exs)
        End If
        Return retval
    End Function

    Shared Function Load(ByRef oPdf As GhostScriptHelper.Pdf, oByteArray() As Byte, maxWidth As Integer, maxHeight As Integer, exs As List(Of Exception)) As Boolean
        Dim oStream As New System.IO.MemoryStream(oByteArray)
        Dim retval As Boolean = Load(oPdf, oStream, maxWidth, maxHeight, exs)
        Return retval
    End Function
    Shared Function LoadToFit(ByRef oPdf As GhostScriptHelper.Pdf, oByteArray() As Byte, maxWidth As Integer, maxHeight As Integer, exs As List(Of Exception)) As Boolean
        Dim oStream As New System.IO.MemoryStream(oByteArray)
        Dim retval As Boolean = LoadToFit(oPdf, oStream, maxWidth, maxHeight, exs)
        Return retval
    End Function

    Shared Function Load(ByRef oPdf As GhostScriptHelper.Pdf, sFilename As String, maxWidth As Integer, maxHeight As Integer, exs As List(Of Exception)) As Boolean
        Dim oStream As System.IO.Stream = New IO.FileStream(sFilename, IO.FileMode.Open, IO.FileAccess.Read)
        Dim retval As Boolean = Load(oPdf, oStream, maxWidth, maxHeight, exs)
        Return retval
    End Function

    Shared Function Load(ByRef oPdf As GhostScriptHelper.Pdf, oStream As System.IO.Stream, maxWidth As Integer, maxHeight As Integer, exs As List(Of Exception)) As Boolean
        Dim retval As Boolean = GhostScriptHelper.Rasterize(oStream, oPdf, exs)
        If retval Then
            If maxWidth = 0 And maxHeight = 0 Then
                oPdf.Thumbnail = oPdf.Portrait
            Else
                oPdf.Thumbnail = LegacyHelper.ImageHelper.GetThumbnailToFill(ImageHelper.FromBytes(oPdf.Portrait), maxWidth, maxHeight).Bytes()
            End If
        End If
        Return retval
    End Function

    Shared Function LoadToFit(ByRef oPdf As GhostScriptHelper.Pdf, oStream As System.IO.Stream, maxWidth As Integer, maxHeight As Integer, exs As List(Of Exception)) As Boolean
        Dim retval As Boolean = GhostScriptHelper.Rasterize(oStream, oPdf, exs)
        If retval Then
            If maxWidth = 0 And maxHeight = 0 Then
                oPdf.Thumbnail = oPdf.Portrait
            Else
                Dim oPortrait = LegacyHelper.ImageHelper.FromBytes(oPdf.Portrait)
                oPdf.Thumbnail = LegacyHelper.ImageHelper.GetThumbnailToFit(oPortrait, maxWidth, maxHeight).Bytes()
            End If
        End If
        Return retval
    End Function

    Shared Function ThumbnailPreview(oDocFile As DTODocFile) As Byte()
        Dim retval As Byte() = Nothing
        If oDocFile IsNot Nothing Then
            Select Case oDocFile.Mime
                Case MimeCods.Bmp, MimeCods.Gif, MimeCods.Jpg, MimeCods.Png, MimeCods.Tif, MimeCods.Tiff
                    Dim oImg = ImageHelper.FromBytes(oDocFile.Thumbnail)
                    retval = ImageHelper.GetThumbnailToFill(oImg, 100, 100).Bytes()
                Case MimeCods.Pdf
                    'retval = LegacyHelper.ImageHelper.Converter(My.Resources.pdf_100)
            End Select
        End If
        Return retval
    End Function

End Class
